firmware/usbpipe.c

```diff
@@ -34,26 +43,61 @@ void usbpipe_disable() {
     bridge_disable_chan(BRIDGE_USB); // Tells SPI Daemon to close USB Daemon
 }
 
-// Received from USB, send to bridge
-void pipe_usb_out_completion() {
-    if (pipe_state_pc_to_soc == PIPE_WAIT_FOR_USB) {
-        u32 len = usb_ep_out_length(USB_EP_PIPE_OUT);
-        bridge_start_in(BRIDGE_USB, pipe_buffer_pc_to_soc, len);
-        pipe_state_pc_to_soc = PIPE_WAIT_FOR_BRIDGE;
-    } else {
-        invalid();
+void out_ring_step() {
+    // If we are not currently sending
+    // And we haven't filled more packets than we're allowed?
+    // And we have less than a full packet?
+    if (!out_usb_pending && out_ring_count < OUT_RING_SIZE && out_ring_short_packet == 0) {
+        // Start reading data in over USB to the buffer at the correct position (all 64 bytes)
+        usb_ep_start_out(USB_EP_PIPE_OUT, out_ring_buf[out_ring_write_pos], PACKET_SIZE);
+        // We are waiting for the transfer to complete
+        out_usb_pending = true;
+    }
+
+    // If we are not waiting on a bridge transaction to complete and we have packets to send
+    if (!out_bridge_pending && out_ring_count > 0) {
+        // The size of the packet is 64 bytes (unless the below case is true)
+        u8 len = PACKET_SIZE;
+        // If we only have one outgoing rick and it is a short packet
+        if (out_ring_count == 1 && out_ring_short_packet != 0) {
+             // The length is actually a subset of a full packet
+             len = out_ring_short_packet;
+             // Reset the out var
+             out_ring_short_packet = 0;
+        }
+        // Start reading in packets from the bridge that should be sent out over USB
+        bridge_start_in(BRIDGE_USB, out_ring_buf[out_ring_write_pos], len);
```

Comment:
And for the bridge, I used the "out" and "in" convention from USB as well, where the SoC is the host, so this is sending from the MCU to the SoC.